### DISCUSSION TITLE
Why NextJS pre-renders all the pages by default?

### DISCUSSION THREAD

**Discussion by pugazhendhis** [vercel/next.js#41191] [Help]
As NextJS pre-renders all the pages from server and it keep maintaining them in server space, which lead high memory usage (Though pages doesn't occupy huge memory, but still i feel it is high memory usage). if application has millions of user base, then the server needs to maintain heap of pages. 

For instance if I have a chat application with millions of user and if User1 had chat with 100s of people then he need to fetch  100s of user chat by individual id, for which server generate 100 pages( for each id fetch ) and it keeps it. same goes to User2 and User3 on and on for all the millions of users using the application. Then finally server will maintains those millions of pages in server which is overburden in server. More over the pages generated for my chat are not require for other millions of users, in this case keeping them in server doesn't make sense.

Please explain me this situation more clearly, Looking forward for your response.
Thanks in advance.

**Comment by icyJoseph** [comment]
Let's look at the case where there's no custom `_app`.

`/` defined as a page without data fetching functions.

```tsx
const Home = () => <div>Home</div>;
export default Home;
```

Then at build time, Next.js pre-renders this page. Even if this page had client side data fetching logic built-in, there'd be only one instance of this page, both HTML and JS on the server.

`/section` defined as a page with `getStaticProps`. At build time, this page is pre-rendered as well using `getStaticProps`' result as props. Three things are generated, HTML, JS and JSON. The HTML contains the initial data for the page as JSON, and the JS contains the instructions on how to make the page interactive, so that it may, for example fetch data client side. The JS and JSON are used when doing SPA transitions, that is done with `next/link`, and `next/router`. It totals up to 3 pages, even if the page had client side data fetching. The HTML is server when user first arrives to the application.

`/data` defined as a page with `getServerSideProps` (GSSP). At build time nothing happens. At runtime, when a request arrives the entire page is server side rendered and send as HTML, JS to enable interactivity follows. During SPA transitions (the user was already on our app), only the result of `getServerSideProps` is send to the user. If prefetching is enabled `next/router` or `next/link` would get the JS for the page ahead of time, but not the JSON.  Otherwise, the JS is fetched on link hover. AFAIK, nothing is really kept on the server after this operation.

Now we have the dynamic paths, and ISR.

`/sub/[foo]` defined as dynamic path without data-fetching functions. It will also be pre-rendered, but the `query` object will be empty. So there'll be only one version of `/sub/[foo]`, which at first render frame must account for an empty query, that is no value of `foo` exists. Once in the client, when the router is ready, Next.js triggers an update with the query value of `foo`.  One page with its HTML and JS is generated.

`/bar/[path]` defined with `getStaticPaths` and `getStaticProps`. This page at build time, is pre-rendered as instructed by `getStaticPaths`, and as many pages as paths are returned will be created. If `getStaticPaths` has `fallback: true`, then one more page, called Fallback is created. At runtime, if `fallback: true` or `fallback: 'blocking'`, then paths not seen at build time will be generated. Here we could potentially generate a page for each number:

```tsx
// pseudo-code, it might not work as is, but the idea does work
const ANumber = (props) => {
 const router = useRouter()
 return router.isFallback ? <div>Loading...</div> : <div>{props.data}</div>
}

export const getStaticPaths = () => {
  return {paths: [], fallback: true}
}

export const getStaticProps = (ctx) => {
 const path = ctx.params?.path;
 // do some checks to make sure path is a number, otherwise return notFound
 return { props: { data: path } }
}
```

This is the only case where you could have an ever increasing number of pages, through ISR (`fallback: true`). 

I am ignoring concepts such as `isrMemoryCacheSize`, `revalidate`, and `on-demand` revalidation.

If you design your chat system so that it has to rely on `getStaticPaths` with fallback set to true, then, without any additional considerations, yeah you'll see a lot of pages being created. 

I'd argue that a chat page between people should be made with GSSP, so that at the moment I click to open the chat, I get the latest list of messages, I can also control server side if you are authenticated and authorised. Alternatively, a skeleton page that fetches the data client side based is also possible, but that's a trade-off. API routes.

For user profile pages, I'd instead pre-render a skeleton page and fill it client side, or again, GSSP, but GSSP is a bit slower and I'd expect my profile to page to show something ASAP.

If I had products, articles, etc, that I am selling, and want to show to the open internet, perhaps I'd use `getStaticPaths` with fallback true, and pre-render the top 1000 most searched articles, and generate the rest on demand. If I had to do this without a framework, I'd have to generate all pages up-front anyway, or I'd wait to do it on a per request basis, like with GSSP.

In all likely-hood we are talking about slightly different situations or concepts, so please do respond. Would love if we can learn something new from each other.

**Comment by yuchant** [comment]
Why does it pre-render all pages by default? 

Because that's its core feature. Next.js is a static site generator first. 

It bridges the common SSG problem of not having a server or per-request features, so it supports SSR and per request operations, but isn't meant to replace a chat server.

The fact it can do anything beyond pre-rendering is one of the things that makes Next.js so attractive. If you wouldn't want a static site generator in the first place, another framework would be useful.

Here's a site on typical Jamstack static generators: https://jamstack.org/generators/ where you'll find similar frameworks of the same type.
